File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/ManualRegistrationScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import java.time.Duration;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class ManualRegistrationScenario {
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private ManualRegistrationScenario() {
    }

    public static void run(ScenarioContext context) {
        Contracts.EchoRes manual = context.client().requestToChannel(
                Contracts.CHANNEL,
                new Contracts.EchoManualReq("manual-request"))
            .packetName("EchoManual")
            .timeout(REQUEST_TIMEOUT)
            .await(Contracts.EchoRes.class);
        ScenarioAssert.ensure("echo:manual-request".equals(manual.value()) && "manual".equals(manual.handler()),
            "RC-A3 request mismatch");
        context.client().sendToChannel(Contracts.CHANNEL, new Contracts.EchoManualMsg("manual-send"))
            .packetName("EchoManual")
            .await();
        ScenarioAssert.waitForEvidence(context.evidence(), "Send", "EchoManual", "manual-send");
        System.out.println("scenario RC-A3 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/ManualRegistrationScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.Contracts
import systems.zlink.e2e.kotlin.registrationcodec.EchoManualMsg
import systems.zlink.e2e.kotlin.registrationcodec.EchoManualReq
import systems.zlink.e2e.kotlin.registrationcodec.EchoManualRes
import systems.zlink.e2e.kotlin.registrationcodec.client.support.EvidenceText
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.framework.channels.ZLinkClient

class ManualRegistrationScenario(
    private val client: ZLinkClient,
    private val evidence: EvidenceText,
    private val assert: ScenarioAssert,
) {
    fun run() {
        val manual = client.requestToChannel(Contracts.CHANNEL, EchoManualReq("manual-request"))
            .packetName("EchoManualReq")
            .timeout(requestTimeout)
            .await(EchoManualRes::class.java)
        assert.that(manual.value == "echo:manual-request" && manual.handler == "manual", "RC-A3 request mismatch")
        client.sendToChannel(Contracts.CHANNEL, EchoManualMsg("manual-send"))
            .packetName("EchoManualMsg")
            .await()
        evidence.waitForEvidence("Send", "EchoManualMsg", "manual-send")
        println("scenario RC-A3 passed")
    }
}
